### PR TITLE
Expose TORCH_XPU_API to Pytorch

### PR DESCRIPTION
**Pull request by etaf**
Expose TORCH_XPU_API to Pytorch

### CODE REVIEW

**Comment by etaf** [comment]
This PR is relied upon stock Pytorch PR: https://github.com/pytorch/pytorch/pull/126678

**Comment by EikanWang** [comment]
@etaf , is it critical to catch up PT 2.4?

**Comment by etaf** [comment]
> @etaf , is it critical to catch up PT 2.4?

No, it's a performance optimization.

**Comment by etaf** [comment]
Since this Pytorch 2.4 release branch has been cut, I think we can start land this PR and the https://github.com/pytorch/pytorch/pull/126678 that depend on it.

File: src/aten/EmptyTensor.h [hunk 1]
```diff
@@ -4,37 +4,38 @@
 namespace at::detail {
 
 // XXX: add TORCH_XPU_API
```
Comment (EikanWang) [inline src/aten/EmptyTensor.h:6]: The comment should be invalid now.

File: src/aten/EmptyTensor.cpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #define TORCH_ASSERT_NO_OPERATORS
 #include <ATen/Context.h>
 #include <ATen/EmptyTensor.h>
+#include <ATen/xpu/EmptyTensor.h>
```
Comment (EikanWang) [inline src/aten/EmptyTensor.cpp:4]: Since we have included the xpu/EmptyTensor, is ATen/EmptyTensor still required?

**Review by EikanWang — COMMENTED** [review]
LGTM.

File: src/aten/EmptyTensor.cpp (hunk 2 above)
Comment (etaf) [reply src/aten/EmptyTensor.cpp]: `ATen/xpu/EmptyTensor.h` is the exported interface of `EmptyTensor.cpp`
and `ATen/EmptyTensor.h` is for `at::detail::empty_generic` which is used in this  `EmptyTensor.cpp` file.

File: src/aten/EmptyTensor.h (hunk 1 above)
Comment (etaf) [reply src/aten/EmptyTensor.h]: removed

File: src/aten/EmptyTensor.cpp (hunk 2 above)
Comment (EikanWang) [reply src/aten/EmptyTensor.cpp]: Then my question will be why `ATen/xpu/EmptyTensor.h` includes `ATen/EmptyTensor.h`?

Comment (etaf) [reply src/aten/EmptyTensor.cpp]: Sorry, but `ATen/xpu/EmptyTensor.h` does not include `ATen/EmptyTensor.h`.

**Comment by etaf** [comment]
@fengyuan14  can you help review this PR?

File: src/aten/EmptyTensor.cpp [hunk 3]
```diff
@@ -1,4 +1,5 @@
 #define TORCH_ASSERT_NO_OPERATORS
+#include "EmptyTensor.h"
```
Comment (fengyuan14) [inline src/aten/EmptyTensor.cpp:2]: Why add it?

Comment (etaf) [reply src/aten/EmptyTensor.cpp:2]: It's the interface of xpu empty_strided.

Comment (fengyuan14) [reply src/aten/EmptyTensor.cpp:2]: Before adding it, we can pass compilation. Why add it?

Comment (fengyuan14) [reply src/aten/EmptyTensor.cpp:2]: Ok, you mean the decorator declared in header need to be included during cpp compilation to let compiler know the attribute.

Comment (fengyuan14) [reply src/aten/EmptyTensor.cpp:2]: Can we use `aten/EmptyTensor.h` here?

Comment (etaf) [reply src/aten/EmptyTensor.cpp:2]: if you mean `#include<aten/EmptyTensor.h>` , yes.

**Comment by fengyuan14** [comment]
You can expose the var to PyTorch cmake,
https://github.com/intel/torch-xpu-ops/blob/a7c3bd2a7f7b17ac692bc7faa3067f3b2a005383/src/CMakeLists.txt#L7
And add it into cmake var of Caffe2 or ATen include dirs.

**Comment by etaf** [comment]
> You can expose the var to PyTorch cmake,
> 
> https://github.com/intel/torch-xpu-ops/blob/a7c3bd2a7f7b17ac692bc7faa3067f3b2a005383/src/CMakeLists.txt#L7
> 
> 
> And append it into cmake var of Caffe2 or ATen include dirs.

Wouldn't that make the two PRs dependent on each other?

**Comment by fengyuan14** [comment]
> > You can expose the var to PyTorch cmake,
> > https://github.com/intel/torch-xpu-ops/blob/a7c3bd2a7f7b17ac692bc7faa3067f3b2a005383/src/CMakeLists.txt#L7
> > 
> > And append it into cmake var of Caffe2 or ATen include dirs.
> 
> Wouldn't that make the two PRs dependent on each other?

ATen_XPU_INCLUDE_DIR is the committed formal and stable API of torch-xpu-ops CMAKE. It is not implementation dependent.

**Comment by fengyuan14** [comment]
Can we do like this ? What do you think of it?
image

**Comment by fengyuan14** [comment]
Looks I have exposed it by CMAKE  target torch-xpu-ops, like this,
image

**Comment by etaf** [comment]
The original design is to align the include path with CPU/CUDA, eg. `ATen/xpu/EmptyTensor.h`, can I still keep this path with your design ?

**Review by fengyuan14 — APPROVED** [review]

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**